### PR TITLE
feat: move instance root volume to another storage pool [WD-14768]

### PR DESCRIPTION
**Pull request by mas-who**
## Done

- for clustered LXD server, introduced a extra step for choosing the migration type between "storage pool" and "cluster member"
- for non-cluster LXD server, clicking the "Migrate" button will only show the storage pool select modal (new component)
- removed the storage pool selector for editing root disk device of an existing instance. The storage pool selector for root disk device will be present if the user is not editing an existing instance.
- also renamed `PanelFormLink` to `FormLink` and generalised its usage. Some of the permission specific logic had to be moved out in to `GroupForm`.
- extracted back linking logic from `GroupHeaderTitle` and generalised its usage into a component

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - In a clustered LXD environment, try migrate the instance to a different cluster member
    - In a clustered LXD environment, try migrate the instance root volume to a different storage pool
    - In a non-clustered LXD environment, try migrate the instance root volume to a different storage pool

## Screenshots

Screenshot from 2024-09-20 15-03-10
Screenshot from 2024-09-20 15-03-40
Screenshot from 2024-09-20 15-03-52

### CODE REVIEW

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:31]:
```suggestion
    ? `Choose migration method`
```

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:41]:
```suggestion
        title={`Choose new ${migrationEntity}`}
```

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:16]:
```suggestion
type MigrationType = "cluster member" | "root storage pool";
```

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:38]:
```suggestion
  if (type !== "") {
```

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:35]:
```suggestion
    ? type
```

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:21]:
```suggestion
  const [type, setType] = useState<MigrationType | null>(isClustered ? null : "root storage pool");
```
Maybe this makes below simpler, we can hide the back link and remove the other special cases?

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:1]: I think we should extract the table to a dedicated component, so we can reuse it in the custom storage volume migration.

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:52]:
```suggestion
      `Root storage migration failed for instance ${instanceName}`,
```

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:79]:
```suggestion
          `Root storage migration started for instance ${instance.name}`,
```

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:44]:
```suggestion
        Instance <ItemName item={{ name: instanceName }} bold /> root storage successfully
        migrated to pool <ItemName item={{ name: newTarget }} bold />
```

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:112]: I think we should move this out of the loop.

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:194]:
```suggestion
        This will migrate the instance <strong>{instance.name}</strong> root storage to
```

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:227]: This should have "go back" semantics

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:74]:
```suggestion
            title="Migrate instance root storage to a different pool"
```

File: src/pages/instances/ClusterMemberMigrationTable.tsx
Comment (mas-who) [reply src/pages/instances/ClusterMemberMigrationTable.tsx:205]: sounds good, I will change the button text to "Go back" as well to be clear :+1:

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (mas-who) [reply src/pages/instances/MigrateInstanceModal.tsx:38]: thanks this is much better! Just need to adjust the `MigrationType` type to include "" as well.

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (mas-who) [reply src/pages/instances/MigrateInstanceModal.tsx:21]: The simplest I can reduce the code below is something like 
```
const modalTitle = !type ? (
    "Choose migration method"
  ) : (
    <>
      {isClustered && (
        <BackLink
          title={target ? "Confirm migration" : `Choose new ${type}`}
          onClick={handleGoBack}
          linkText={target ? `Choose ${type}` : "Choose migration method"}
        />
      )}
      {!isClustered && (target ? "Confirm migration" : `Choose new ${type}`)}
    </>
  );
```
Is this what you had in mind? it's less code but I think it's still a little messy

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:1]: will tackle this one, once #916 is merged

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:112]: good catch thanks!

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (mas-who) [reply src/pages/instances/MigrateInstanceModal.tsx:74]: I double check the modal heading, summary copy and notification copy. It's consistently referring to "root storage" now.

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:1]: @edlerd I added an extra commit moving logic out of `StoragePoolMigrationTable` and `ClusterMemberMigrationTable` into a custom hook `useInstanceMigration`. This should allow the table components to be reused easily for custom storage volume migration. The actual implementation might change a little bit depending on the final state for #916, let me know what you think?

Comment (edlerd) [reply src/pages/instances/StoragePoolMigrationTable.tsx:1]: I am not sure if you need a hook for this. My idea was a pure rendering component for the pool selector. Removing the instance, and actions from the select logic:
```
const StoragePoolTableSelect: FC<Props> = ({
  onSelect,
}) => {
```

So the component only loads and renders the pools as a table with a select button next to it. It is purely to select a pool. You might need to pass in some info to disable the current row, but no more things should go into it. It can load the pools internally, render them and call a parent on click.

Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:1]: right, this makes sense to me too. I created the `useInstanceMigration` custom hook for the reason that there is a fair bit of duplicated logic between `StoragePoolMigrationTable` and `ClusterMemberMigrationTable`. So the goal is a bit more than just splitting out the visual aspects of the storage pool selection table

Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:1]: @edlerd made another change which splits out the storage pool selection into its own component `StoragePoolSelectTable`. I kept the custom hook `useInstanceMigration` for now though since I think it's quite nice to combine the repetitive stateful logic between cluster member and storage pool migration for instances.

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [inline src/pages/instances/MigrateInstanceModal.tsx:41]:
```suggestion
      close();
```
I think this should move to the bottom of the function and we don't need the comment, because we can also cancel on the 1st step in the clustered case and want to close under that condition.

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:46]:
```suggestion
```
We should load the data, when it is needed, not unconditionally in this hook.

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:121]: I don't think we should call close on failure, probably only in the then block. Also I'd rename it to `onSuccess`, wdyt?

File: src/pages/instances/ClusterMemberMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/ClusterMemberMigrationTable.tsx:12]:
```suggestion
```
We should load the members in here and add a loading indicator.

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:34]: This state shouldn't live in the hook, we should pass in target and type. But the state should live in the parent caller make the interface of the hook smaller.

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:132]:
```suggestion
    handleMigrate,
```

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:24]:
```suggestion
export const useInstanceMigration = ({ instance, close, type, target }: Props) => {
```
We tend to destructure the props in the header everywhere else in the repo.

File: src/util/instanceMigration.tsx
Comment (edlerd) [inline src/util/instanceMigration.tsx:20]:
```suggestion
  close: () => void;
  type: MigrationType;
  target: string;
```

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolMigrationTable.tsx:19]:
```suggestion
const InstanceStoragePoolMigration: FC<Props> = ({
```
Not a table anymore, and I think we should be more specific to avoid the name clash with the volume pool migration.

File: src/pages/instances/StoragePoolSelectTable.tsx
Comment (edlerd) [inline src/pages/instances/StoragePoolSelectTable.tsx:1]: This should probably live in the storage module and not under instances.

File: src/components/forms/DiskDeviceFormRoot.tsx
Comment (mas-who) [reply src/components/forms/DiskDeviceFormRoot.tsx:125]: sure, we still need to conditionally apply `u-no-margin--bottom` though, otherwise the help text will appear right against the selector's bottom border.

Comment (edlerd) [reply src/components/forms/DiskDeviceFormRoot.tsx:125]: > sure, we still need to conditionally apply `u-no-margin--bottom` though, otherwise the help text will appear right against the selector's bottom border.

Yes, that is why I kept in the suggestion above :)

File: src/pages/instances/StoragePoolSelectTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolSelectTable.tsx:10]: that was my initial approach, however since this is a more general component, we should consider what if we want to disable multiple rows?  We could pass in a list of objects with your suggested structure, but I think it would be easier to allow the definition of a disable function since through the callback arguments we can get all the info we need for a storage pool wdyt?

File: src/components/forms/DiskDeviceFormRoot.tsx
Comment (mas-who) [reply src/components/forms/DiskDeviceFormRoot.tsx:125]: not seeing it :sweat_smile: `className` is assigned with "u-no-margin--bottom" without condition?

Comment (edlerd) [reply src/components/forms/DiskDeviceFormRoot.tsx:125]: I am confused, you are probably right :)

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (mas-who) [reply src/pages/instances/MigrateInstanceModal.tsx:41]: `migrationType` would be set to "root storage pool" initially if lxd is not clustered. Moving `close` to the bottom would result in the modal not being able to close.

File: src/pages/instances/StoragePoolSelectTable.tsx
Comment (edlerd) [reply src/pages/instances/StoragePoolSelectTable.tsx:10]: I was thinking about this too and as for the two use cases at hand we have only one row to disable, probably fine to only support that case for now. If we ever need more, we can enhance it.

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (edlerd) [reply src/pages/instances/MigrateInstanceModal.tsx:41]: Oh right, there is the special case for legacy. I don't really like that it increases complexity here, but don't have a better idea. Probably fine as is.

Do we need another `close()` at the end of the function for the first step in the clustered case?

File: src/pages/instances/StoragePoolSelectTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolSelectTable.tsx:10]: cool, done :+1:

File: src/pages/instances/MigrateInstanceModal.tsx
Comment (mas-who) [reply src/pages/instances/MigrateInstanceModal.tsx:41]: The first step for the clustered case does not have a cancel button. The modal at that point will be closed by `close()` always.

File: src/util/instanceMigration.tsx
Comment (mas-who) [reply src/util/instanceMigration.tsx:46]: 100%, I wasn't quite done with this part yet since I was planning on waiting for #916 to get merged. tbh I don't think we need to wait anymore though, I'm quite sure that `StoragePoolSelectTable` can be quite easily implemented in a follow up PR

Comment (mas-who) [reply src/util/instanceMigration.tsx:46]: these calls will be moved into the respective table components

**Comment by mas-who** [comment]
@edlerd one more change I'd like to make is to also split out the select table for cluster members. The code would be more consistent that way and also if we do want to add the feature for migrating custom volume to another cluster member later, we can reuse the table component.

File: src/util/instanceMigration.tsx
Comment (mas-who) [reply src/util/instanceMigration.tsx:121]: yeah that sounds good :+1:

File: src/util/instanceMigration.tsx
Comment (mas-who) [reply src/util/instanceMigration.tsx:34]: yes that's neat! :+1:

File: src/pages/instances/StoragePoolMigrationTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolMigrationTable.tsx:19]: same goes for `ClusterMemberMigrationTable` -> `InstanceClusterMemberMigration`

File: src/pages/instances/StoragePoolSelectTable.tsx
Comment (mas-who) [reply src/pages/instances/StoragePoolSelectTable.tsx:1]: yes, same goes for `ClusterMemberSelectTable`, moving it to `/cluster`

File: src/util/instanceMigration.tsx
Comment (edlerd) [reply src/util/instanceMigration.tsx]: The handle and on prefix is correct, I meant renaming the close prop (which doesn't fit any of the two oO) to onSuccess.

File: src/components/forms/DiskDeviceFormRoot.tsx
Comment (edlerd) [inline src/components/forms/DiskDeviceFormRoot.tsx:121]: We need the class also in the non-edit/create case

image

File: src/pages/storage/StoragePoolSelectTable.tsx
Comment (edlerd) [inline src/pages/storage/StoragePoolSelectTable.tsx:46]: I think we should find a better way to visualize the disabled row. Maybe remove the hover effect for it, or we could try muted text in that row. Or hide it instead of disabling?

**Review by edlerd — COMMENTED** [review]
This looks way cleaner, thanks for the changes. Two more comments below.

File: src/util/instanceMigration.tsx
Comment (mas-who) [reply src/util/instanceMigration.tsx]: oh my bad for misunderstanding, yes that makes sense :+1:

File: src/components/forms/DiskDeviceFormRoot.tsx
Comment (mas-who) [reply src/components/forms/DiskDeviceFormRoot.tsx:121]: aah yes, actually we need to hide the help text if we are not editing an instance. `u-no-margin--bottom` is only needed when there is no help text in order to reduce vertical spacing between the pool selector and the size input

File: src/pages/storage/StoragePoolSelectTable.tsx
Comment (mas-who) [reply src/pages/storage/StoragePoolSelectTable.tsx:46]: going with muted text, show not-allowed cursor for a disabled row. wdyt?

Comment (edlerd) [reply src/pages/storage/StoragePoolSelectTable.tsx:46]: Sounds OK.
The not allowed cursor is barely used on the web, maybe just remove the hover effect and use the default pointer?

**Review by edlerd — DISMISSED** [review]
LGTM

**Comment by mas-who** [comment]
> LGTM

Thanks for the review! I'm quite happy about how clean the code has become :) Time for a design review please @piperdeck :pray:

**Comment by piperdeck** [comment]
This modal has a primary CTA button at the bottom which is always disabled, and never able to be clicked. Could we simply remove the entire CTA footer from the modal, since the "Cancel" functionality is already provided by the close button at the top right?

CleanShot 2024-09-24 at 16 53 19@2x

I've also noticed that the modals for moving custom storage volumes and instance root volumes have different columns, different title formats, and slightly different dimensions. Could we make these modals match each other?


| Custom volume                                                                                                     | Instance                                                                                                      |
| ----------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------- |
| CleanShot 2024-09-24 at 16 55 05@2x | CleanShot 2024-09-24 at 16 55 00@2x |

For the title, I would go with:

| Custom volume                                       | Instance                                           |
| --------------------------------------------------- | -------------------------------------------------- |
| Choose storage pool for custom volume `volume name` | Choose storage pool for instance `<instance name>` |

and for the columns, I would keep the one being used by the root volume modal, and show those columns in that order for the instance modal as well.

**Comment by mas-who** [comment]
> This modal has a primary CTA button at the bottom which is always disabled, and never able to be clicked. Could we simply remove the entire CTA footer from the modal, since the "Cancel" functionality is already provided by the close button at the top right?
> 
> CleanShot 2024-09-24 at 16 53 19@2x
> 

Yeah agreed, I will remove the action buttons for non-clustered scenario. In a clustered environment, the cancel button leads to prior steps in the modal flow.

> I've also noticed that the modals for moving custom storage volumes and instance root volumes have different columns, different title formats, and slightly different dimensions. Could we make these modals match each other?
> 
> Custom volume	Instance
> CleanShot 2024-09-24 at 16 55 05@2x	CleanShot 2024-09-24 at 16 55 00@2x
> For the title, I would go with:
> 
> Custom volume	Instance
> Choose storage pool for custom volume `volume name`	Choose storage pool for instance `<instance name>`
> and for the columns, I would keep the one being used by the root volume modal, and show those columns in that order for the instance modal as well.

@piperdeck agreed on both :) nice one catching the difference 

> This modal has a primary CTA button at the bottom which is always disabled, and never able to be clicked. Could we simply remove the entire CTA footer from the modal, since the "Cancel" functionality is already provided by the close button at the top right?
> 
> CleanShot 2024-09-24 at 16 53 19@2x
> 
> I've also noticed that the modals for moving custom storage volumes and instance root volumes have different columns, different title formats, and slightly different dimensions. Could we make these modals match each other?
> 
> Custom volume	Instance
> CleanShot 2024-09-24 at 16 55 05@2x	CleanShot 2024-09-24 at 16 55 00@2x
> For the title, I would go with:
> 
> Custom volume	Instance
> Choose storage pool for custom volume `volume name`	Choose storage pool for instance `<instance name>`

Sure I can update the copy text for the modal title. @edlerd this is contradictory to what's been changed in the code review, although I do agree with @piperdeck that having the instance/custom volume names here provides context to the resource that we are modifying, which is better IMO even if it makes the title a little longer. We should probably truncate the instance/volume name in the modal title though.
 
> and for the columns, I would keep the one being used by the root volume modal, and show those columns in that order for the instance modal as well.

@piperdeck do you mean keep the columns in the custom volume migration modal and use them in the instance root volume migration modal?

**Comment by mas-who** [comment]
> > This modal has a primary CTA button at the bottom which is always disabled, and never able to be clicked. Could we simply remove the entire CTA footer from the modal, since the "Cancel" functionality is already provided by the close button at the top right?
> > CleanShot 2024-09-24 at 16 53 19@2x
> 
> Yeah agreed, I will remove the action buttons for non-clustered scenario. In a clustered environment, the cancel button leads to prior steps in the modal flow.
> 
> > I've also noticed that the modals for moving custom storage volumes and instance root volumes have different columns, different title formats, and slightly different dimensions. Could we make these modals match each other?
> > Custom volume	Instance
> > CleanShot 2024-09-24 at 16 55 05@2x	CleanShot 2024-09-24 at 16 55 00@2x
> > For the title, I would go with:
> > Custom volume	Instance
> > Choose storage pool for custom volume `volume name`	Choose storage pool for instance `<instance name>`
> > and for the columns, I would keep the one being used by the root volume modal, and show those columns in that order for the instance modal as well.
> 
> > I've also noticed that the modals for moving custom storage volumes and instance root volumes have different columns, different title formats, and slightly different dimensions. Could we make these modals match each other?

Yes we will make the modal consistent, these are two separate PRs for now though.

> > For the title, I would go with:
> > Custom volume	Instance
> > Choose storage pool for custom volume `volume name`	Choose storage pool for instance `<instance name>`
> 
> Sure I can update the copy text for the modal title. @edlerd this is contradictory to what's been changed in the code review, although I do agree with @piperdeck that having the instance/custom volume names here provides context to the resource that we are modifying, which is better IMO even if it makes the title a little longer. We should probably truncate the instance/volume name in the modal title though.
> 
> > and for the columns, I would keep the one being used by the root volume modal, and show those columns in that order for the instance modal as well.
> 
> @piperdeck do you mean keep the columns in the custom volume migration modal and use them in the instance root volume migration modal?

**Comment by piperdeck** [comment]
+1 on the modal title.

Everything looks good to me!

**Comment by mas-who** [comment]
@edlerd I have replaced the custom volume storage pool selection table with the `StoragePoolSelectTable` component in this PR. If all is good, would you approve again thanks :)

**Comment by mas-who** [comment]
@edlerd I've also added a e2e test for this feature, relating to this [task](https://warthogs.atlassian.net/jira/software/c/projects/WD/boards/930?assignee=712020%3Accb7b61f-3f6a-41ec-9935-a79519dada4e&selectedIssue=WD-15104). Note the test checks if the lxd environment is clustered since the instance root storage migration steps are different and we always have a lxd cluster in CI.

File: src/components/BackLink.tsx
Comment (edlerd) [reply src/components/BackLink.tsx:10]: That is not a good solution. Also the truncate doesn't work on the selector.

I think we should not have any custom CSS on the title and also remove the truncation. The heading will naturally wrap to multiple lines then.

Comment (mas-who) [reply src/components/BackLink.tsx:10]: perfect, thanks that makes things a whole lot simpler :+1:

**Review by edlerd — APPROVED** [review]
QA and Code LGTM, good work :+1:

**Pull request merged**
